clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapChainStorageEngine.java

```diff
@@ -0,0 +1,624 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.clustered.server.offheap;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.locks.Lock;
+import org.ehcache.clustered.common.store.Chain;
+import org.ehcache.clustered.common.store.Element;
+import org.ehcache.clustered.common.store.SequencedElement;
+import org.ehcache.clustered.common.store.Util;
+import org.terracotta.offheapstore.paging.OffHeapStorageArea;
+import org.terracotta.offheapstore.paging.PageSource;
+import org.terracotta.offheapstore.storage.PointerSize;
+import org.terracotta.offheapstore.storage.StorageEngine;
+import org.terracotta.offheapstore.storage.portability.Portability;
+
+import static java.util.Collections.unmodifiableList;
+
+class OffHeapChainStorageEngine<K> implements StorageEngine<K, InternalChain> {
+
+  private static final int ELEMENT_HEADER_SEQUENCE_OFFSET = 0;
+  private static final int ELEMENT_HEADER_LENGTH_OFFSET = 8;
+  private static final int ELEMENT_HEADER_NEXT_OFFSET = 12;
+  private static final int ELEMENT_HEADER_SIZE = 20;
+
+  private static final int CHAIN_HEADER_KEY_LENGTH_OFFSET = 0;
+  private static final int CHAIN_HEADER_KEY_HASH_OFFSET = 4;
+  private static final int CHAIN_HEADER_TAIL_OFFSET = 8;
+  private static final int CHAIN_HEADER_SIZE = 16;
+
+  private final OffHeapStorageArea storage;
+  private final Portability<? super K> keyPortability;
+  private final Set<AttachedInternalChain> activeChains = new HashSet<AttachedInternalChain>();
+
+  private StorageEngine.Owner owner;
+  private long nextSequenceNumber = 0;
+
+  public OffHeapChainStorageEngine(PageSource source, Portability<? super K> keyPortability, boolean thief, boolean victim) {
+    this.storage = new OffHeapStorageArea(PointerSize.LONG, new StorageOwner(), source, 4096, thief, victim);
```

Comment:
No particular reason.  I'll expose these further out, and switch to more reasonable defaults, but I think at least for now defaulting inside the offheap specific code makes sense.